cmake/MirCommon.cmake

```diff
@@ -126,13 +126,34 @@ function (mir_discover_tests_internal EXECUTABLE TEST_ENV_OPTIONS DETECT_FD_LEAK
     set(test_env ${env_without_preloads})
   endif()
 
-  # Final commands
-  set(test_cmd "${test_cmd}" "--gtest_filter=-${test_no_memcheck_filter}:${test_exclusion_filter}")
-  set(test_cmd_no_memcheck "${test_cmd_no_memcheck}" "--gtest_death_test_style=threadsafe" "--gtest_filter=${test_no_memcheck_filter}:-${test_exclusion_filter}")
   if(DETECT_FD_LEAKS)
     set(test_cmd ${CMAKE_SOURCE_DIR}/tools/detect_fd_leaks.bash ${test_cmd})
   endif()
 
+  # ptest
+  list(JOIN test_env " --env " discover_env)
+  string(PREPEND discover_env "--env ")
+
+  list(JOIN test_cmd " " discover_cmd)
+  list(JOIN test_cmd_no_memcheck " " discover_cmd_no_memcheck)
+
+  list(JOIN test_exclusion_filter ":" test_exclusion_str)
+  list(JOIN test_no_memcheck_filter ":" memcheck_exclusion_str)
+
+  if (test_no_memcheck_filter)
+    file(APPEND ${CMAKE_BINARY_DIR}/discover_all_tests.sh
+      "sh ${CMAKE_SOURCE_DIR}/tools/discover_gtests.sh ${discover_env} --test-name ${test_name} --gtest-exclude ${test_exclusion_str} --gtest-exclude ${memcheck_exclusion_str} -- ${discover_cmd}\n")
+    file(APPEND ${CMAKE_BINARY_DIR}/discover_all_tests.sh
+      "sh ${CMAKE_SOURCE_DIR}/tools/discover_gtests.sh ${discover_env} --test-name ${test_name}_no_memcheck --gtest-exclude ${test_exclusion_str} --gtest-include ${memcheck_exclusion_str} -- ${discover_cmd_no_memcheck}\n")
+  else()
+    file(APPEND ${CMAKE_BINARY_DIR}/discover_all_tests.sh
+      "sh ${CMAKE_SOURCE_DIR}/tools/discover_gtests.sh ${discover_env} --test-name ${test_name} --gtest-exclude ${test_exclusion_str} -- ${discover_cmd}\n")
+  endif()
+
+  # Final commands
+  set(test_cmd "${test_cmd}" "--gtest_filter=-${memcheck_exclusion_str}:${test_exclusion_str}")
+  set(test_cmd_no_memcheck "${test_cmd_no_memcheck}" "--gtest_death_test_style=threadsafe" "--gtest_filter=${memcheck_exclusion_str}:-${test_exclusion_str}")
```

Comment:
Since you key by `test_no_memcheck_filter` above, could move those up there to avoid memcheck exclusions if they're empty? Not a strong voice, mind you, since if they're empty it'll just mean an extra leading `:`, but…